tests/tracer/test_propagation.py

```diff
@@ -20,19 +20,216 @@
+    """When we have a single large tag that won't fit"""
+    # DEV: Limit is 128 for key=value
+    meta = {"_dd.p.dm": ("x" * 128)[len("_dd.p.dm") - 1 :]}
+    ctx = Context(trace_id=1234, sampling_priority=2, dd_origin="synthetics", meta=meta)
+    tracer.context_provider.activate(ctx)
+    with tracer.trace("global_root_span") as span:
+        headers = {}
+        HTTPPropagator.inject(span.context, headers)
+
+        assert _HTTP_HEADER_TAGS not in headers
+        assert ctx._meta["_dd.propagation_error"] == "max_size"
+
+
+def test_inject_tags_invalid(tracer):
+    # DEV: "=" and "," are not allowed in keys or values
+    meta = {"_dd.p.dm": ",value="}
+    ctx = Context(trace_id=1234, sampling_priority=2, dd_origin="synthetics", meta=meta)
+    tracer.context_provider.activate(ctx)
+    with tracer.trace("global_root_span") as span:
+        headers = {}
+        HTTPPropagator.inject(span.context, headers)
+
+        assert _HTTP_HEADER_TAGS not in headers
+        assert ctx._meta["_dd.propagation_error"] == "encoding_error"
+
+
+def test_inject_tags_disabled(tracer):
+    with override_global_config(dict(_x_datadog_tags_enabled=False)):
+        meta = {"_dd.p.dm": "value"}
+        ctx = Context(trace_id=1234, sampling_priority=2, dd_origin="synthetics", meta=meta)
+        tracer.context_provider.activate(ctx)
+        with tracer.trace("global_root_span") as span:
+            headers = {}
+            HTTPPropagator.inject(span.context, headers)
+
+            assert _HTTP_HEADER_TAGS not in headers
+            assert ctx._meta["_dd.propagation_error"] == "disabled"
+
+
+def test_inject_tags_previous_error(tracer):
+    """When we have previously gotten an error, do not try to propagate tags"""
+    # This value is valid
+    meta = {"_dd.p.dm": "value", "_dd.propagation_error": "some fake test value"}
+    ctx = Context(trace_id=1234, sampling_priority=2, dd_origin="synthetics", meta=meta)
+    tracer.context_provider.activate(ctx)
+    with tracer.trace("global_root_span") as span:
+        headers = {}
+        HTTPPropagator.inject(span.context, headers)
+
+        assert _HTTP_HEADER_TAGS not in headers
+
+
+def test_extract(tracer):
+    headers = {
+        "x-datadog-trace-id": "1234",
+        "x-datadog-parent-id": "5678",
+        "x-datadog-sampling-priority": "1",
+        "x-datadog-origin": "synthetics",
+        "x-datadog-tags": "_dd.p.dm=value,any=tag",
+    }
+
+    context = HTTPPropagator.extract(headers)
+
+    tracer.context_provider.activate(context)
+
+    with tracer.trace("local_root_span") as span:
+        assert span.trace_id == 1234
+        assert span.parent_id == 5678
+        assert span.context.sampling_priority == 1
+        assert span.context.dd_origin == "synthetics"
+        assert span.context._meta == {
+            "_dd.origin": "synthetics",
+            "_dd.p.dm": "value",
+            "any": "tag",
+        }
+        # TODO: Should x-datadog-tags be in shallow copy?
```

Comment:
@Kyle-Verhoog In adding a multispan trace as you suggested, we see that the activated context (with its `_meta`) is shallow copied when a child span is started. I had assumed these "trace level" metadata would not be copied to all spans but I realize now that it would have to given how we inject in the propagators using the context on the current span rather than the local root span. Should we retain this behavior for `x-datadog-tags`?

Reply:
Still need to add snapshot test for this. to integration test.